cirq-core/cirq/devices/device.py

```diff
@@ -253,3 +263,108 @@ def _from_json_dict_(cls, qubits, nx_graph, **kwargs):
         if nx_graph != '':
             graph_obj = nx.readwrite.json_graph.node_link_graph(nx_graph)
         return cls(qubits, graph_obj)
+
+
+@value.value_equality
+class GridDeviceMetadata(DeviceMetadata):
+    """Hardware metadata for homogenous 2d symmetric grid devices."""
+
+    def __init__(
+        self,
+        qubit_pairs: Iterable[Tuple['cirq.Qid', 'cirq.Qid']],
+        supported_gates: 'cirq.Gateset',
+        gate_durations: Optional[Dict['cirq.Gateset', 'cirq.Duration']] = None,
+    ):
+        """Create a GridDeviceMetadata object.
+
+        Create a GridDevice which has a well defined set of couplable
+        qubit pairs that have the same two qubit gates available in
+        both coupling directions.
+
+        Args:
+            qubit_pairs: Iterable of pairs of `cirq.Qid`s representing
+                bi-directional couplings.
+            supported_gates: `cirq.Gateset` indicating gates supported
+                everywhere on the device.
+            gate_durations: Optional dictionary of `cirq.Gateset`
+                instances mapping to `cirq.Duration` instances for
+                gate timing metadata information. If provided,
+                must match all entries in supported_gates.
+
+        Raises:
+            ValueError: if the union of gateset keys in gate_durations,
+                do not represent an identical gateset to supported_gates.
+        """
+        qubit_pairs = list(qubit_pairs)
+        flat_pairs = [q for pair in qubit_pairs for q in pair]
+        # Keep lexigraphically smaller tuples for undirected edges.
+        sorted_pairs = sorted(qubit_pairs)
+        qubit_set = set()
+        for a, b in sorted_pairs:
+            if (b, a) not in qubit_set:
+                qubit_set.add((a, b))
+
+        connectivity = nx.Graph()
+        connectivity.add_edges_from(sorted(list(qubit_set)), directed=False)
+        super().__init__(flat_pairs, connectivity)
+        self._qubit_pairs = frozenset(qubit_set)
+        self._supported_gates = supported_gates
+
+        if gate_durations is not None:
+            working_gatefamilies = frozenset(
+                g for gset in gate_durations.keys() for g in gset.gates
+            )
+            if working_gatefamilies != supported_gates.gates:
+                overlap = working_gatefamilies.difference(supported_gates.gates)
```

Comment:
Nit: overlap would be elements in both sets, right?

Reply:
Fixed variable name.